Go deadlock update

jrajahalme opened the pull request:
Updating the github.com/sasha-s/go-deadlock dependency to the current master with the vendor tool adds detection of duplicate locking. The diff on the README.md is like this:

```
 A.Lock() // defer A.Unlock() or similar.
+
+Another common sources of deadlocs is duplicate take a lock in a goroutine:
+
+A.Rlock() or lock()
+
+A.lock() or A.RLock()
+
+
 This does not guarantee a deadlock (maybe the goroutines above can never be running at the same time), but it usually a design flaw at least.
```
The problem is that we have seemed to assume it top be safe for the same goroutine to take an RLock twice. In reality, if another goroutine takes a write Lock on the same lock in between the two RLocks in the first goroutine, both of these goroutines will block forever (deadlock). This is due to go RW lock being "write preferred", meaning that after a write lock is blocking, all further read lock attempts block., and the first read lock will never be released.

Even the pkg/lock unit test fails with this updated go-deadlock, as it also assumes that taking a read lock twice in the same goroutine should succeed.

Fix this and a detected double RLock in the option package.

Fixes: #2205 
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on pkg/option/option.go:
```diff
@@ -169,13 +169,17 @@ func NewBoolOptions(lib *OptionLibrary) *BoolOptions {
 
-func (bo *BoolOptions) IsEnabled(key string) bool {
-	bo.optsMU.RLock()
+func (bo *BoolOptions) IsEnabledLocked(key string) bool {
```
set this a private method. `s/IsEnabledLocked/isEnabledLocked`

aanm commented on pkg/option/option.go:
```diff
@@ -197,7 +201,9 @@ func (bo *BoolOptions) SetIfUnset(key string, value bool) {
 
 func (bo *BoolOptions) InheritDefault(parent *BoolOptions, key string) {
-	bo.Set(key, parent.IsEnabled(key))
+	bo.optsMU.Lock()
```
`s/Lock/RLock`

aanm commented on pkg/option/option.go:
```diff
@@ -197,7 +201,9 @@ func (bo *BoolOptions) SetIfUnset(key string, value bool) {
-	bo.Set(key, parent.IsEnabled(key))
+	bo.optsMU.Lock()
+	bo.Opts[key] = parent.IsEnabledLocked(key)
+	bo.optsMU.Unlock()
```
`s/Unlock/RUnlock`

jrajahalme replied:
Done.

aanm commented on pkg/option/option.go:
```diff
@@ -234,15 +240,15 @@ func ParseOption(arg string, lib *OptionLibrary) (string, bool, error) {
+// getFmtOptLocked returns #define name if option exists and is set to true in endpoint's Opts
 // map or #undef name if option does not exist or exists but is set to false
-func (bo *BoolOptions) getFmtOpt(name string) string {
+func (bo *BoolOptions) getFmtOptLocked(name string) string {
```
IMO I don't think it's necessary to write "Locked" on private methods.

private methods -> lock already acquired.
public methods -> call from outside so we need to acquire lock.

This is just a thought, what do you think?

jrajahalme replied:
By this logic private "isEnabledLocked" should then be called "isEnabled", where the locking difference to public "IsEnabled" should be deduced from the name. Not sure if this is what you want?

jrajahalme replied:
Of maybe you meant that "Locked" suffix is not needed on private methods when there is no corresponding public function?

I'm going by this and doing exactly what you asked for :-)

aanm replied:
>Of maybe you meant that "Locked" suffix is not needed on private methods when there is no corresponding public function?

Correct

jrajahalme committed:
pkg: Remove double RLocking.

Updating the github.com/sasha-s/go-deadlock dependency to the current
master with the vendor tool adds detection of duplicate locking.

The problem is that we have seemed to assume it top be safe for the
same goroutine to take an RLock twice. In reality, if another
goroutine takes a write Lock on the same lock in between the two
RLocks in the first goroutine, both of these goroutines will block
forever (deadlock). This is due to go RW lock being "write preferred",
meaning that after a write lock is blocking, all further read lock
attempts block., and the first read lock will never be released.

Even the pkg/lock unit test fails with this updated go-deadlock, as
it also assumes that taking a read lock twice in the same goroutine
should succeed.

Fix this and a detected double RLock in the option package.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
go-deadlock: Update to current master for double lock detection.

Updating the github.com/sasha-s/go-deadlock dependency to the current master with the vendor tool adds detection of duplicate locking. The diff on the README.md is like this:

 A.Lock() // defer A.Unlock() or similar.
+
+Another common sources of deadlocs is duplicate take a lock in a goroutine:
+
+A.Rlock() or lock()
+
+A.lock() or A.RLock()
+
+
 This does not guarantee a deadlock (maybe the goroutines above can never be running at the same time), but it usually a design flaw at least.


Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/option/option.go:
```diff
@@ -115,6 +115,9 @@ func (om OptionMap) DeepCopy() OptionMap {
 
+// BoolOptions member functions with external access do not require
+// locking by the caller, while functions with internal access presume
+// the caller to have taken care of any locking needed.
```
I've seen a style before in golang that names functions `fooLocked()`, `fooRLocked()` to indicate that they should be called while holding the lock. Might be a bit weighty in this case, but something to consider.

joestringer replied:
I see this was already discussed above ;) Feel free to disregard.

joestringer reviewed (approved):
LGTM
